util/helm/client.go

```diff
@@ -100,62 +100,62 @@ func (c *nativeHelmChart) ExtractChart(chart string, version *semver.Version) (s
 
 	exists, err := fileExist(chartPath)
 	if err != nil {
-		return "", nil, err
+		return "", "", nil, err
 	}
 	if !exists {
 		helmCmd, err := NewCmd(c.repoPath)
 		if err != nil {
-			return "", nil, err
+			return "", "", nil, err
 		}
 		defer helmCmd.Close()
 
 		_, err = helmCmd.Init()
 		if err != nil {
-			return "", nil, err
+			return "", "", nil, err
 		}
 
 		_, err = helmCmd.RepoUpdate()
 		if err != nil {
-			return "", nil, err
+			return "", "", nil, err
 		}
 
 		// (1) because `helm fetch` downloads an arbitrary file name, we download to an empty temp directory
 		tempDest, err := ioutil.TempDir("", "helm")
 		if err != nil {
-			return "", nil, err
+			return "", "", nil, err
 		}
 		defer func() { _ = os.RemoveAll(tempDest) }()
 		_, err = helmCmd.Fetch(c.repoURL, chart, version.String(), tempDest, c.creds)
 		if err != nil {
-			return "", nil, err
+			return "", "", nil, err
 		}
 		// (2) then we assume that the only file downloaded into the directory is the tgz file
 		// and we move that to where we want it
 		infos, err := ioutil.ReadDir(tempDest)
 		if err != nil {
-			return "", nil, err
+			return "", "", nil, err
 		}
 		if len(infos) != 1 {
-			return "", nil, fmt.Errorf("expected 1 file, found %v", len(infos))
+			return "", "", nil, fmt.Errorf("expected 1 file, found %v", len(infos))
 		}
 		err = os.Rename(filepath.Join(tempDest, infos[0].Name()), chartPath)
 		if err != nil {
-			return "", nil, err
+			return "", "", nil, err
 		}
 	}
 	// untar helm chart into throw away temp directory which should be deleted as soon as no longer needed
 	tempDir, err := ioutil.TempDir("", "helm")
 	if err != nil {
-		return "", nil, err
+		return "", "", nil, err
 	}
 	cmd := exec.Command("tar", "-zxvf", chartPath)
 	cmd.Dir = tempDir
 	_, err = executil.Run(cmd)
 	if err != nil {
 		_ = os.RemoveAll(tempDir)
-		return "", nil, err
+		return "", "", nil, err
 	}
-	return path.Join(tempDir, chart), util.NewCloser(func() error {
+	return path.Join(tempDir, chart), tempDir, util.NewCloser(func() error {
```

Comment:
I think in case of a helm chart root directory is just a helm chart directory itself. So `ExtractChart` should not return root dir and we can use helm chart path as root directory

Reply:
Do you mean that we should use the chart itself as the root directory?

If, for example, we use `/foo/bar/Chart.yaml` as the root directory, under the current path-traversal logic a user wouldn't be able to use a values file even under the same directory. Consider this example:

```
foo/bar/
├── Chart.yaml
├── templates
│   ├── NOTES.txt
│   ├── _helpers.tpl
│   ├── deployment.yaml
│   ├── ingress.yaml
│   ├── service.yaml
│   └── tests
│       └── test-connection.yaml
└── values.yaml
```

In this case a user wouldn't be able to use `values.yaml`.

Currently we treat the directory where the `Chart.yaml` as the "root directory" (in this example `/foo/bar`), in that case anything under the root directory is fair game.

I don't have much experience with Helm repos, so sorry if I'm missing something obvious.